### PR TITLE
[CARBONDATA-1856][PARTITION] Support insert/load data for partition table

### PR DESCRIPTION
**Pull request by ravipesala**
This PR depends on PR https://github.com/apache/carbondata/pull/1642

Changed carbonrelation to HadoopFSRelation during load in case of the partition table. 
Implement sparks Fileformat interface for carbon and use carbonoutputformat inside.
Create partition.map file inside each segment for mapping between partition and index file.

The following syntax is supported now.
```
CREATE TABLE partitiontwo (empno int, designation String,
  workgroupcategory int, workgroupcategoryname String, deptno int, deptname String,
  projectcode int, projectjoindate Timestamp, projectenddate Timestamp,attendance int,
  utilization int,salary int)
 PARTITIONED BY (doj Timestamp, empname String)
 STORED BY 'org.apache.carbondata.format'

LOAD DATA local inpath '$resourcesPath/data.csv' INTO TABLE partitiontwo OPTIONS('DELIMITER'= ',', 'QUOTECHAR'= '"')

insert into insertpartitionthree select empno,doj,workgroupcategoryname,deptno,deptname,projectcode,projectjoindate,projectenddate,attendance,utilization,salary,workgroupcategory,empname,designation from originTable

LOAD DATA local inpath '$resourcesPath/data.csv' INTO TABLE singlepasspartitionone OPTIONS('DELIMITER'= ',', 'QUOTECHAR'= '"', 'SINGLE_PASS'='true')

```

Be sure to do all of the following checklist to help us incorporate 
your contribution quickly and easily:

 - [ ] Any interfaces changed?
 
 - [ ] Any backward compatibility impacted?
 
 - [ ] Document update required?

 - [ ] Testing done
        Please provide details on 
        - Whether new unit test cases have been added or why no new tests are required?
        - How it is tested? Please attach test report.
        - Is it a performance related change? Please attach the performance test report.
        - Any additional information to help reviewers in testing this change.
       
 - [ ] For large changes, please consider breaking it into sub-tasks under an umbrella JIRA.

### CODE REVIEW

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.0, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/700/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/1929/

**Comment by ravipesala** [comment]
SDV Build Success , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/2262/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/1939/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.0, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/713/

**Comment by ravipesala** [comment]
SDV Build Fail , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/2268/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.0, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/731/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/1960/

**Comment by ravipesala** [comment]
SDV Build Success , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/2282/

**Comment by jackylk** [comment]
@ravipesala This PR implements Hive dynamic partition. Load into static partition is still pending?

File: core/src/main/java/org/apache/carbondata/core/metadata/PartitionFileStore.java [hunk 1]
```diff
@@ -0,0 +1,199 @@
...
+import org.apache.carbondata.core.fileoperations.FileWriteOperation;
+import org.apache.carbondata.core.util.CarbonUtil;
+import org.apache.carbondata.core.util.path.CarbonTablePath;
+
+import com.google.gson.Gson;
+
+/**
+ * Reads and writes the partition mapper.
+ */
+public class PartitionFileStore {
```
Comment (jackylk) [inline core/src/main/java/org/apache/carbondata/core/metadata/PartitionFileStore.java]: Suggest to change to `PartitionMapFileStore`

File: core/src/main/java/org/apache/carbondata/core/metadata/PartitionFileStore.java [hunk 2]
```diff
@@ -0,0 +1,199 @@
...
+import org.apache.carbondata.core.fileoperations.AtomicFileOperations;
+import org.apache.carbondata.core.fileoperations.AtomicFileOperationsImpl;
+import org.apache.carbondata.core.fileoperations.FileWriteOperation;
+import org.apache.carbondata.core.util.CarbonUtil;
+import org.apache.carbondata.core.util.path.CarbonTablePath;
+
+import com.google.gson.Gson;
+
+/**
+ * Reads and writes the partition mapper.
```
Comment (jackylk) [inline core/src/main/java/org/apache/carbondata/core/metadata/PartitionFileStore.java]: Suggest to change to `Provide read and write support for partition mapping file in each segment`

File: core/src/main/java/org/apache/carbondata/core/metadata/schema/partition/PartitionType.java [hunk 3]
```diff
@@ -23,5 +23,6 @@
   RANGE,
   RANGE_INTERVAL,
   LIST,
-  HASH
+  HASH,
+  STANDARD
```
Comment (jackylk) [inline core/src/main/java/org/apache/carbondata/core/metadata/schema/partition/PartitionType.java]: Maybe `HIVE` is better

File: core/src/main/java/org/apache/carbondata/core/metadata/schema/table/CarbonTable.java [hunk 4]
```diff
@@ -573,6 +574,11 @@ public boolean isPartitionTable() {
     return null != tablePartitionMap.get(getTableName());
   }
 
+  public boolean isStandardPartitionTable() {
```
Comment (jackylk) [inline core/src/main/java/org/apache/carbondata/core/metadata/schema/table/CarbonTable.java]: change to `isHivePartitionTable`

**Comment by ravipesala** [comment]
@jackylk This PR implements both static and dynamic partition, but the syntax for static partition through LOAD command is not done yet. We can INSERT INTO sql for static partition. Test case also added for it please check.

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.0, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/808/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/2039/

**Comment by ravipesala** [comment]
SDV Build Fail , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/2352/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.0, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/831/

File: core/src/main/java/org/apache/carbondata/core/metadata/PartitionFileStore.java (hunk 1 above)
Comment (ravipesala) [reply core/src/main/java/org/apache/carbondata/core/metadata/PartitionFileStore.java]: ok

File: core/src/main/java/org/apache/carbondata/core/metadata/PartitionFileStore.java (hunk 2 above)
Comment (ravipesala) [reply core/src/main/java/org/apache/carbondata/core/metadata/PartitionFileStore.java]: ok

File: core/src/main/java/org/apache/carbondata/core/metadata/schema/partition/PartitionType.java (hunk 3 above)
Comment (ravipesala) [reply core/src/main/java/org/apache/carbondata/core/metadata/schema/partition/PartitionType.java]: ok

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/2058/

File: core/src/main/java/org/apache/carbondata/core/metadata/schema/table/CarbonTable.java (hunk 4 above)
Comment (ravipesala) [reply core/src/main/java/org/apache/carbondata/core/metadata/schema/table/CarbonTable.java]: ok

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.0, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/836/

**Comment by ravipesala** [comment]
SDV Build Success , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/2367/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/2062/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.0, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/841/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/2067/

**Comment by ravipesala** [comment]
SDV Build Fail , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/2372/

**Comment by ravipesala** [comment]
retest this please

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.0, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/854/

**Comment by ravipesala** [comment]
retest this please

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.0, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/855/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/2080/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/2082/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.0, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/857/

**Comment by ravipesala** [comment]
SDV Build Success , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/2380/

**Comment by jackylk** [comment]
LGTM

File: integration/spark2/src/main/scala/org/apache/spark/sql/execution/datasources/CarbonFileFormat.scala [hunk 5]
```diff
@@ -0,0 +1,247 @@
...
+    conf.set("carbon.commit.protocol", "carbon.commit.protocol")
+    sparkSession.sessionState.conf.setConfString(
+      "spark.sql.sources.commitProtocolClass",
+      "org.apache.spark.sql.execution.datasources.CarbonSQLHadoopMapReduceCommitProtocol")
+
+    job.setOutputFormatClass(classOf[CarbonTableOutputFormat])
+
+    var table = CarbonMetadata.getInstance().getCarbonTable(
+      options.getOrElse("dbName", "default"), options("tableName"))
+//    table = CarbonTable.buildFromTableInfo(table.getTableInfo, true)
```
Comment (jackylk) [inline integration/spark2/src/main/scala/org/apache/spark/sql/execution/datasources/CarbonFileFormat.scala]: remove it

File: integration/spark2/src/main/scala/org/apache/spark/sql/execution/datasources/CarbonFileFormat.scala [hunk 6]
```diff
@@ -0,0 +1,247 @@
...
+      classOf[CarbonOutputCommitter],
+      classOf[CarbonOutputCommitter])
+    conf.set("carbon.commit.protocol", "carbon.commit.protocol")
+    sparkSession.sessionState.conf.setConfString(
+      "spark.sql.sources.commitProtocolClass",
+      "org.apache.spark.sql.execution.datasources.CarbonSQLHadoopMapReduceCommitProtocol")
+
+    job.setOutputFormatClass(classOf[CarbonTableOutputFormat])
+
+    var table = CarbonMetadata.getInstance().getCarbonTable(
```
Comment (jackylk) [inline integration/spark2/src/main/scala/org/apache/spark/sql/execution/datasources/CarbonFileFormat.scala]: Can you use CarbonEnv.getCarbonTable

File: integration/spark2/src/main/scala/org/apache/spark/sql/execution/datasources/CarbonFileFormat.scala [hunk 7]
```diff
@@ -0,0 +1,247 @@
...
+
+  override def shortName(): String = "carbondata"
+
+  override def inferSchema(sparkSession: SparkSession,
+      options: Map[String, String],
+      files: Seq[FileStatus]): Option[StructType] = {
+    None
+  }
+
+  override def prepareWrite(sparkSession: SparkSession,
```
Comment (jackylk) [inline integration/spark2/src/main/scala/org/apache/spark/sql/execution/datasources/CarbonFileFormat.scala]: move parameter to next line

File: integration/spark2/src/main/scala/org/apache/spark/sql/execution/datasources/CarbonFileFormat.scala (hunk 5 above)
Comment (ravipesala) [reply integration/spark2/src/main/scala/org/apache/spark/sql/execution/datasources/CarbonFileFormat.scala]: ok

File: integration/spark2/src/main/scala/org/apache/spark/sql/execution/datasources/CarbonFileFormat.scala (hunk 6 above)
Comment (ravipesala) [reply integration/spark2/src/main/scala/org/apache/spark/sql/execution/datasources/CarbonFileFormat.scala]: ok

File: integration/spark2/src/main/scala/org/apache/spark/sql/execution/datasources/CarbonFileFormat.scala (hunk 7 above)
Comment (ravipesala) [reply integration/spark2/src/main/scala/org/apache/spark/sql/execution/datasources/CarbonFileFormat.scala]: ok

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.0, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/866/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/2091/

**Comment by ravipesala** [comment]
retest this please

**Comment by jackylk** [comment]
retest this please

**Comment by ravipesala** [comment]
SDV Build Fail , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/2389/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.0, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/874/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/2099/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.0, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/877/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/2102/

File: hadoop/src/main/java/org/apache/carbondata/hadoop/api/CarbonTableOutputFormat.java [hunk 8]
```diff
@@ -339,28 +345,47 @@ private static void setFileHeader(Configuration configuration, CarbonLoadModel m
...
 
-    @Override
-    public void close(TaskAttemptContext taskAttemptContext) {
+    @Override public void close(TaskAttemptContext taskAttemptContext) throws InterruptedException {
       iteratorWrapper.close();
-      dataLoadExecutor.close();
+      try {
+        future.get();
+      } catch (ExecutionException e) {
+        throw new InterruptedException(e.getMessage());
```
Comment (xubo245) [inline hadoop/src/main/java/org/apache/carbondata/hadoop/api/CarbonTableOutputFormat.java]: Why did you throw InterruptedException?

**Pull request closed without merging**